File: core/src/main/java/pl/alpheratzteam/obfuscator/transformer/optimizers/GotoInlinerTransformer.java
package pl.alpheratzteam.obfuscator.transformer.optimizers;

import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.JumpInsnNode;
import pl.alpheratzteam.obfuscator.Obfuscator;
import pl.alpheratzteam.obfuscator.transformer.Transformer;
import java.util.Arrays;
import java.util.Map;
import java.util.Objects;

public class GotoInlinerTransformer extends Transformer
{
    public GotoInlinerTransformer(Obfuscator obfuscator) {
        super(obfuscator);
    }

    @Override
    public void visit(Map<String, ClassNode> classMap) {
        classMap.values().forEach(classNode ->
                classNode.methods.forEach(methodNode ->
                        Arrays.stream(methodNode.instructions.toArray())
                                .filter(abstractInsnNode -> abstractInsnNode.getOpcode() == GOTO)
                                .forEach(abstractInsnNode -> {
                                    final JumpInsnNode gotoJump = (JumpInsnNode) abstractInsnNode;
                                    final AbstractInsnNode insnAfterTarget = gotoJump.label.getNext();
                                    if (Objects.isNull(insnAfterTarget))
                                        return;

                                    if (insnAfterTarget.getOpcode() != GOTO)
                                        return;

                                    gotoJump.label = ((JumpInsnNode) insnAfterTarget).label;
                                })
                )
        );
    }
}

File: src/main/kotlin/pl/alpheratzteam/obfuscator/transformer/optimizer/GotoInlinerTransformer.kt
package pl.alpheratzteam.obfuscator.transformer.optimizer

import org.objectweb.asm.Opcodes.GOTO
import org.objectweb.asm.tree.JumpInsnNode
import pl.alpheratzteam.obfuscator.Obfuscator
import pl.alpheratzteam.obfuscator.api.transformer.Transformer
import pl.alpheratzteam.obfuscator.util.ConditionUtil
import java.util.*

/**
 * @author Unix
 * @since 18.12.2020
 */

class GotoInlinerTransformer : Transformer {

    override fun transform(obfuscator: Obfuscator) {
        obfuscator.classes.values.forEach {
            it.methods.forEach {
                it.instructions.filter { it.opcode == GOTO }.forEach {
                    val gotoJump = it as JumpInsnNode
                    val insnAfterTarget = gotoJump.label.next
                    ConditionUtil.checkCondition(Objects.nonNull(insnAfterTarget)) {
                        when (insnAfterTarget.opcode) {
                            GOTO -> gotoJump.label = (insnAfterTarget as JumpInsnNode).label
                        }
                    }
                }
            }
        }
    }

}